---
layout: post
title: nifH database for dada2
categories: [database]
tags: [amplicon, microbiome]
---

I have some nifH gene amplicon sequencing data and I have never dealt with them before. Here are some road blocks I encountered.

1. How to trim the sequence.

There are currently two ways in my dada2 pipeline, the ITS way where the lengths of the genes vary a lot so that I only set the minimum length based on previous knowledge and suggestions. Then there is the 16S way where the length of the fragments are pretty much fixed and you just need to see where the quality of the sequence drops and truncate there. What's the case for nifH?

After removing reads containing Ns and primer removal, the quality profile of one of my samples looks like this:

![plotQualityProfile](https://user-images.githubusercontent.com/7167719/200302139-21328bc0-fb6a-4993-96aa-157af866accf.png) 

There seems to be some variable starting from 50bp to 250bp, but 90% of the reads makes it to 250bp. Let's take the 16S route but setting the truncLen as 250bp.

2. Then comes the Annotation.

I googled nifH database. The first thing I found was from the [Buckley Lab](https://blogs.cornell.edu/buckley/nifh-sequence-database/). There is a fasta version but it dates 2012.

The second thing I found was from the [Zehr Lab](https://wwwzehr.pmc.ucsc.edu/nifH_Database_Public/). It is quite up-to-date (April 2021). It comes with a fasta file and a metadata file. The fasta file does not have the full rank of each entry but we could fix that based on the accession #. The metadata file is not necessary in the classification but very informative:

		acc,GI,full_name,author,clone,collected_by,collection_date,country,date,description,full_taxon_name,gene,host,isolation_source,lat_lon,mol_type,note,nuc_acc_and_pos,title,sequence_source,AMINO_July2012
		
It came with the suffix of .nds. I do not know what that stands for but it is basically a csv. The only catch is that some of the columns are wrapped with quotes and some are not. That is something to be fixed later.

Before working on retrieving the full ranks of my fasta file, I would like to know whether I am re-inventing the wheel.

From the papers I've read so far, people usually use the OTU pipeline and the nifH has its own arb database (under the same link from the Zehr Lab). 

I chose one paper to read with care: 

[Evaluation of Primers Targeting the Diazotroph Functional Gene and Development of NifMAP â€“ A Bioinformatics Pipeline for Analyzing nifH Amplicon Data](https://www.frontiersin.org/articles/10.3389/fmicb.2018.00703/full)

Yes I still read papers from __frontiers in microbiology__ even though one of my NSFC reviewer thinks that I should not pushlish there.

They studied 8 primers (4F and 4R) but the pair I used (Soares 2006) was not included. Umm... OK, can I still use their NifMAP pipeline? Their [resources](https://github.com/roey-angel/NifMAP/tree/master/Resources) seems a bit out of date.

What else.

Oh, a [2022 New Phytologiest paper](https://nph.onlinelibrary.wiley.com/doi/abs/10.1111/nph.17837)! Oh well they did qPCR so only the abundance was measured, not diversity.

Oh my, look what I found! A [Bioinformatics paper](https://academic.oup.com/bioinformatics/article/37/16/2289/6134496) with a [R package](https://github.com/bkapili/ppit)! Looks promising.

 




		