---
layout: post
title: FAPROTAX
categories: [analysis]
tags: [amplicon, microbiome]
---

## Download and install

The current version is [1.2.6](https://pages.uoregon.edu/slouca/LoucaLab/archive/FAPROTAX/SECTION_Download/MODULE_Downloads/CLASS_Latest%20release/UNIT_FAPROTAX_1.2.6/FAPROTAX_1.2.6.zip). There is a very good [tutorial for 1.1.1 in Chinese](https://blog.csdn.net/woodcorpse/article/details/106552599) from which I learnt a lot.

No installation is needed. The whole package is basically a python script plus a text format data file.

It is also very easy to use:

./collapse_table.py -i otu_table.biom -o functional_table.biom -g FAPROTAX.tx
t

The input file is different from [PICRUST2](http://fanhuan.github.io/en/2022/07/05/Picrust2-basic/). This one needs the taxonomy annotation, similar to the one I received from Rhonin. How do I generate one? I need to write a function, from a phyloseq object to this format.

	~/build/FAPROTAX_1.2.6/collapse_table.py -i 16S_fastq_mean_aerial_allthree_merged_TMM_7449tips.otu_table.txt -o functional_table.biom -g ~/build/FAPROTAX_1.2.6/FAPROTAX.txt 
	WARNING: The module 'biom' was not found.
	         This script will only accept classical (e.g. tab-separated) tables
	WARNING: The module 'h5py' was not found.
	         This script may not work with tables in BIOM HDF5 format
	Traceback (most recent call last):
	  File "/home/hfan/build/FAPROTAX_1.2.6/collapse_table.py", line 1892, in <module>
	    biom_table = biom.table.Table(	collapsed_table_numerical_part, 
	NameError: name 'biom' is not defined

	$ pip install Cython
	$ pip install biom-format
	
	

