---
layout: post
title: Is Population Genetics Important For Quantitative Genetics?
categories: [Opinions]
tags: [PopGen][QuantGen]
---

If one looks at my publication list, they would think I am a phylogeny person. Yes, I have devoted some good years in phylogenetics, a.k.a. reconstructing evolutionary relationships between species. However along the way (since I stayed long enough in universities and research institutes), I also had some training in population genetics and quantitative genetics. In fact, the first course I took in my PhD program was Genetics 629 Evolutionary Biology. It had three modules. The first one is **Phylogenetics** taught by Prof. Bret Larget from Statistics. This is totally my comfort zone and I already know most of the things he was teaching (I had a master before this). 

The second module is **Population Genetics**, taught by Prof. Bret Payseur from Genetics. I've had Population Genetics a couple of times by then. Firstly in my undergrad in Ecology. Yes this might sound a bit weird. We had Genetics, but we talked about it more from a molecular point of view instead of a population point of view (later I know it would be microevolution). Therefore the population ecology part is the closest I got about population genetics as a undergrad. Then during my masters, I had two courses about Evolution, one called micro-evolution and one called macro-evolution. I remember being still confused at the end of the semester for what is micro and what is macro and why we need to have this distinction. The real breakthrough came during the month-long Quantitative Genetics workshop in Okinawa, where I learned not only the basic concepts, but also the hotly debated topics back then. The organizer purposefully invited speakers with opposite views on the same topic and put them back to back so they can argue with each other. It was really fun to watch since I was so removed from those research that I do not need to pick a side. Recombination vs. drift to name just one. 

I did not know any Bret before Gen629, and suddenly I know two, and they have completely different personality. I still remember I was gone for a field course and missed a test for Payseur's module so he had to give me a makeup test. When we were riding the lift up to my test room, I asked whether I could as a question. Maybe it was already too much trouble for him to arrange this makeup test just for me, he was a bit annoied and asked me "why". I kept my mouth shut until the end of his module but it had been very helpful. It was the first time that I hand-calculated those important parameters such as effective population size. Hand calculation is the only way to fully understand a numeric concept, which I believe till this day. 

The third module comes the **Quantitative Genetics** given by Prof. John Doebley. Of course I've never heard him, or Payseur before, but somehow I vaguely sensed that he is a big deal. John was extremely nice, but did not like to answer some questions on the spot, at least my friend Lauren would say so. We were so naive about this subject and can ask very stupid questions sometimes. I really liked this module even though this is the first time I was exposed to this category of knowledge, partly still due to the hand calculation for things such as narrow and broad sense heritability. I was also fascinated by the possibility of linking genotypes with phenotypes, which is by far the most important research question to me. Later I took more courses on this topic and learned about GWAS, and eventually this became the third chapter of my PhD thesis.

Now my work has nothing to do with phylogenetics and is almost solely about quantitative genetics, I always wonders where Population Genetics stand in my work. Today when I was trying to filter out some SNPs so I do not need to deal with dozens of millions of them, I gave the Hardy-Weinburg Evaluation another thought. It is often desirable to filter out loci based on statistically significant (for a given Î±-value or P value) deviations from H-W proportions. Everybody knows about Hardy-Weinburg, the iconic p+q=1 -> p^2 + pq + q^2 = 1, simple, elegant, yet too good to be true, just like effective population size (random mating, infinitely large population in the absence of selection, migration, or new mutation). Therefore the concern is that a deviation from it might not be a consequence of poor SNP quality. If we filter too aggressively, we might lose the ones that are actually interesting since their allele frequency might be different in different populations.

Then it strikes. How much population genetics knowledge do I need for quantitative genetics? 

They ["both focus on the genetic basis of phenotypic variation among  the individuals in a population. Population genetics traditionally deals with frequencies of alleles and genotypes, whereas quantitative  genetics concentrates on the ways that individual variation in genotype  and environment contribute to variance in phenotype."](https://www.nature.com/scitable/topic/population-and-quantitative-genetics-21/)

So, they share some of the basic concepts but focusing on different problems. One more theoretical and one more practical. Maybe later I can say more about how knowledge from population genetics can help. 





